tests/zfs-tests/tests/functional/dnode_stress/dnode_stress.ksh

```diff
@@ -0,0 +1,80 @@
+#!/bin/ksh -p
+#
+# CDDL HEADER START
+#
+# The contents of this file are subject to the terms of the
+# Common Development and Distribution License (the "License").
+# You may not use this file except in compliance with the License.
+#
+# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+# or https://opensource.org/licenses/CDDL-1.0.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+# Stress the dirty dnode bug:
+#     issue - https://github.com/openzfs/zfs/issues/15526
+#     fix - https://github.com/openzfs/zfs/pull/15571)
+#
+# Copyright (c) 2023 by Lawrence Livermore National Security, LLC.
+
+. $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/tests/functional/reservation/reservation.shlib
+. $STF_SUITE/tests/functional/zvol/zvol_common.shlib
+
+verify_runnable "global"
+
+function reproducer {
+	cd /$TESTPOOL
+	mypid=$(exec sh -c 'echo $PPID')
+
+	prefix="reproducer_${mypid}_"
+	log_must dd if=/dev/urandom of=${prefix}0 bs=1M count=1 status=none
+
+	log_note "writing files"
+	end=1000
```

Comment:
The worst error rate I've seen is around 8 errors per million files. With four workers and 1000 files each, that's around a 3.2% chance of losing the race for one of these test runs, which in turn is an expected value of 20ish runs to get a true negative. 10 000 files takes a few minutes on a beefy system, which may be too much for the test suite, but perhaps this can be bumped up a bit.